src/order/filter/at_top_bot.lean

```diff
@@ -1571,6 +1571,33 @@ lemma tendsto_of_seq_tendsto {f : α → β} {k : filter α} {l : filter β} [k.
   (∀ x : ℕ → α, tendsto x at_top k → tendsto (f ∘ x) at_top l) → tendsto f k l :=
 tendsto_iff_seq_tendsto.2
 
+lemma tendsto_inf_principal_iff_seq_tendsto {f : α → β} {k : filter α} {l : filter β}
+  (s : set α) [(k ⊓ 𝓟 s).is_countably_generated] :
+  tendsto f (k ⊓ 𝓟 s) l
+    ↔ (∀ x : ℕ → α, (∀ n, x n ∈ s) → tendsto x at_top k → tendsto (f ∘ x) at_top l) :=
+begin
+  have hs := eq_empty_or_nonempty s,
+  cases hs with hs hs,
+  { simp only [hs, principal_empty, inf_bot_eq, tendsto_bot, mem_empty_iff_false,
+      is_empty.forall_iff, implies_true_iff], },
+  rw tendsto_iff_seq_tendsto,
+  simp_rw [tendsto_inf, tendsto_principal],
+  refine ⟨λ h x hx_mem hx_tendsto, h x ⟨hx_tendsto, eventually_of_forall hx_mem⟩,
+    λ h x hx, _⟩,
+  obtain ⟨x0, hx0s⟩ := hs,
+  let y : ℕ → α := λ n, if x n ∈ s then x n else x0,
+  have hy_eq_x : y =ᶠ[at_top] x,
+  { filter_upwards [hx.2] with n hxsn_mem,
+    simp_rw [y, if_pos hxsn_mem], },
+  refine (tendsto_congr' _).mp (h y _ ((tendsto_congr' hy_eq_x).mpr hx.1)),
+  { filter_upwards [hy_eq_x] with n hn,
+    rw [function.comp_apply, hn], },
+  { intros n,
+    simp_rw y,
+    split_ifs with h' h',
+    exacts [h', hx0s], },
+end
```

Comment:
You can prove it in a few lines
```suggestion
calc tendsto f (k ⊓ 𝓟 s) l ↔ tendsto (f ∘ coe : s → β) (comap coe (k ⊓ 𝓟 s)) l :
  by rw [← @tendsto_map'_iff _ _ _ f coe, map_comap, subtype.range_coe, inf_assoc, inf_idem]
... ↔ ∀ x : ℕ → s, tendsto x at_top (comap coe (k ⊓ 𝓟 s)) → tendsto (f ∘ coe ∘ x) at_top l :
  tendsto_iff_seq_tendsto
... ↔ ∀ x : ℕ → α, (∀ n, x n ∈ s) → tendsto x at_top k → tendsto (f ∘ x) at_top l :
  (equiv.pi_subtype_equiv (λ x y, y ∈ s)).symm.surjective.forall.trans $ by simp [(∘)]
```

if you add
```lean
@[simps apply symm_apply]
def pi_subtype_equiv {π : α → Sort*} (p : Π a, π a → Prop) :
  (Π a, {y // p a y}) ≃ {f : Π a, π a // ∀ a, p a (f a)} :=
{ to_fun := λ f, ⟨λ a, f a, λ a, (f a).2⟩,
  inv_fun := λ f a, ⟨f.1 a, f.2 a⟩,
  left_inv := λ f, funext $ λ a, subtype.ext rfl,
  right_inv := λ ⟨f, hf⟩, rfl }
```
(do we have this equivalence under another name? I failed to find it)